### PR TITLE
React wrapper: Rewrite HotTable & HotColumn into functional components

### PR DESCRIPTION
**Pull request by NOtherDev**
### Context

Getting rid of the class-based approach in favor of functional components.

Basically, internal structure change, but I'm marking it as a breaking change because if someone had used `ref` on `HotTable` they used to get the component class instance while now the dedicated `HotTableRef` interface is provided. Also, previously exported `HotTableClass` is now replaced with `HotTableInner`, although as there is no reason for it to be public anymore, it is not exported.

### How has this been tested?

* all the unit tests still pass
* all the demo projects still run correctly

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature or improvement (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Additional language file or change to the existing one (translations)

### Related issue(s):
1. Proposal in [RFC](https://github.com/handsontable/handsontable/discussions/10767)

### Affected project(s):
- [ ] `handsontable`
- [ ] `@handsontable/angular`
- [x] `@handsontable/react`
- [ ] `@handsontable/vue`
- [ ] `@handsontable/vue3`

### Checklist:

- [x] I have reviewed the guidelines about [Contributing to Handsontable](https://github.com/handsontable/handsontable/blob/master/CONTRIBUTING.md) and I confirm that my code follows the code style of this project.
- [x] I have signed the [Contributor License Agreement](https://docs.google.com/forms/d/e/1FAIpQLScpMq4swMelvw3-onxC8Jl29m0fVp5hpf7d1yQVklqVjGjWGA/viewform?c=0&w=1)
- [ ] My change requires a change to the documentation.

### CODE REVIEW

File: wrappers/react/src/hotTableInner.tsx [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+  useUpdateEffect
+} from './helpers';
+import PropTypes from 'prop-types';
+import { getRenderer } from 'handsontable/renderers/registry';
+import { getEditor } from 'handsontable/editors/registry';
+import { useHotTableContext } from './hotTableContext'
+import { HotColumnContextProvider } from './hotColumnContext'
+import { EditorContextProvider, makeEditorClass } from "./hotEditor";
+
+const HotTableInner = React.forwardRef<HotTableRef, HotTableProps>((props, ref) => {
```
Comment (NOtherDev) [inline wrappers/react/src/hotTableInner.tsx:24]: I initially planned to merge this (what used to be `HotTableClass` into `HotTable` itself, but it's easier to use context if we keep this separation - `HotTable` is the context provider while `HotTableInner` is its consumer. I'm open for the better inner component name suggestion.

Comment (evanSe) [reply wrappers/react/src/hotTableInner.tsx:24]: Your justification for having a distinct `hotTable.tsx` and `hotTableInner.tsx` makes sense to me, I wouldn't worry about changing the `HotTableInner` as its pretty self-explanatory and not exposed to the end user.

File: wrappers/react/src/hotColumn.tsx [hunk 2]
```diff
@@ -6,129 +6,87 @@ import {
...
 const internalProps = ['_columnIndex', '_getOwnerDocument', 'children'];
 
-interface HotColumnInnerProps extends HotColumnProps {
-  _columnIndex: number;
-  _getOwnerDocument: () => Document | null;
-}
-
-class HotColumnInner extends React.Component<HotColumnInnerProps, {}> {
-  columnSettings: Handsontable.ColumnSettings;
+const HotColumn: React.FC<HotColumnProps> = (props) => {
```
Comment (evanSe) [inline wrappers/react/src/hotColumn.tsx:17]: A welcome change 😄

**Review by evanSe — APPROVED** [review]
Great stuff 👏🏻

**Pull request merged**